node/network/dispute-distribution/src/receiver/peer_queues.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{hash_map::Entry, HashMap, VecDeque};
+
+use futures::future::pending;
+use futures_timer::Delay;
+use polkadot_node_network_protocol::request_response::{v1::DisputeRequest, IncomingRequest};
+use polkadot_primitives::v2::AuthorityDiscoveryId;
+
+use crate::RECEIVE_RATE_LIMIT;
+
+/// How many messages we are willing to queue per peer (validator).
+///
+/// The larger this value is, the larger bursts are allowed to be without us dropping messages. On
+/// the flip side this gets allocated per validator, so for a size of 10 this will result
+/// in 10_000 * size_of(`IncomingRequest`) in the worst case.
+///
+/// `PEER_QUEUE_CAPACITY` must not be 0 for obvious reasons.
+#[cfg(not(test))]
+pub const PEER_QUEUE_CAPACITY: usize = 10;
+#[cfg(test)]
+pub const PEER_QUEUE_CAPACITY: usize = 2;
+
+/// Queues for messages from authority peers for rate limiting.
+///
+/// Invariants ensured:
+///
+/// 1. No queue will ever have more than `PEER_QUEUE_CAPACITY` elements.
+/// 2. There are no empty queues. Whenever a queue gets empty, it is removed. This way checking
+///    whether there are any messages queued is cheap.
+/// 3. As long as not empty, `pop_reqs` will, if called in sequence, not return `Ready` more often
+///    than once for every `RECEIVE_RATE_LIMIT`, but it will always return Ready eventually.
+/// 4. If empty `pop_reqs` will never return `Ready`, but will always be `Pending`.
+pub struct PeerQueues {
+	/// Actual queues.
+	queues: HashMap<AuthorityDiscoveryId, VecDeque<IncomingRequest<DisputeRequest>>>,
```

Comment:
how do we handle key rotations?

Reply:
Then there will be two entries at session boundaries, which [should be fine](https://github.com/paritytech/polkadot/blob/465f2667b6a80bad283906a5ab9c60670e615247/node/network/dispute-distribution/src/receiver/mod.rs#L280-L282). This is at least true if validators are prevented from flip flopping their authority ids in the DHT. Otherwise they would be able to bypass the rate limit. :thinking: .... should not be possible as refreshing the DHT takes like 10 minutes.